homeassistant/components/file_upload/__init__.py

```diff
@@ -146,18 +149,34 @@ async def _upload_file(self, request: web.Request) -> web.Response:
         file_upload_data: FileUploadData = hass.data[DOMAIN]
         file_dir = file_upload_data.file_dir(file_id)
 
-        def _sync_work() -> None:
+        def _sync_create_dir() -> None:
             file_dir.mkdir()
 
-            # MyPy forgets about the isinstance check because we're in a function scope
-            assert isinstance(file_field, web.FileField)
-
-            with (file_dir / file_field.filename).open("wb") as target_fileobj:
-                shutil.copyfileobj(file_field.file, target_fileobj)
-
-        await hass.async_add_executor_job(_sync_work)
-
-        file_upload_data.files[file_id] = file_field.filename
+        await hass.async_add_executor_job(_sync_create_dir)
+
+        def _sync_write_file(
+            _file_name: str, _chunk: bytes, _append_to_file: bool
+        ) -> None:
+            mode = "ab" if _append_to_file else "wb"
+            with (file_dir / _file_name).open(mode) as target_fileobj:
```

Comment:
Open the file in the executor and keep it open. Then do each write in the executor. Make sure you close the file handle if anything fails in a finally

Reply:
Thanks! Implemented it like that and it actually makes everything look much cleaner.